src/libraries/System.Private.CoreLib/src/System/Environment.Unix.cs

```diff
@@ -112,5 +112,11 @@ private static unsafe bool TryGetUserNameFromPasswd(byte* buf, int bufLen, out s
         }
 
         private static int GetCurrentProcessId() => Interop.Sys.GetPid();
+
+        /// <summary>
+        /// Returns the path of the executable that started the currently executing process. Returns null when the path is not available.
+        /// </summary>
+        /// <returns>Path of the executable that started the currently executing process</returns>
+        public static string? ProcessPath => Interop.Sys.GetProcessPath();
```

Comment:
What is the benefit of not caching this value, as done for ProcessId?

Reply:
ProcessId is known to be called frequently on hot paths. This API is unlikely to be called frequently on hot paths. Having said that, it may be ok to cache the value. I would like to figure out the desirable behavior first.

Reply:
The interesting corner-cases are:

1. Change current directory, call the API for the first time. Should it be guaranteed to work?
2. Rename the executable, call the API for the first time. Should it return the original executable path or the new executable path?
3. Call the API, rename the executable. call the API again. Should it return the original executable path or the new executable path?

Reply:
Some evidence which supports that returning the original executable path is reasonable:

1.
    * dotnet (5.0 RC2 single file): throws InvalidOperationException when trying to access current process' StartInfo
    * go: returns the old path from `os.Executable()` after doing `os.Chdir("/tmp")`
2.
    * dotnet (5.0 RC2 single file): throws FileNotFoundException from `File.Move()` `Process.GetCurrentProcess().Process.MainModule.FileName`
    * go: returns the old path from `os.Executable()` after doing `os.Rename(os.Executable(), "/tmp/blah")`
3.
    * dotnet (5.0 RC2 single file): returns the old path from `Process.GetCurrentProcess().Process.MainModule.FileName`
    * go: returns the old path from `os.Executable()` before and after doing `os.Rename(os.Executable(), "/tmp/blah")`

ps:
case 2 in dotnet 5.0 RC2 single file seems unexpected; this code does not throw `FileNotFoundException` if we uncomment the fourth line:

```c#
static void Case2()
{
    using Process process = Process.GetCurrentProcess();
    // Console.WriteLine(process.MainModule.FileName);
    File.Move(process.MainModule.FileName, "/tmp/foo"); // throws FNFE. uncomment line above and it will not throw
    Console.WriteLine(process.MainModule.FileName);
}

/* Exception:
Unhandled exception. System.IO.FileNotFoundException: Could not load file or assembly 'System.Threading, Version=5.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'. The system cannot find the file specified.

File name: 'System.Threading, Version=5.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'
   at System.Console.<get_Out>g__EnsureInitialized|25_0()
   at System.Console.get_Out()
   at System.Console.WriteLine(String value)
   at Program.Case2()
   at Program.Main()
Abort trap: 6
*/

// deployed with:              dotnet publish -c Release -p:PublishSingleFile=True -r osx-x64
// ran with absolute path:     /Users/am11/projects/exepath1/bin/Release/net5.0/osx-x64/publish/exepath1
```

Reply:
Thanks for validating the behavior with golang.

The .NET 5 single file has number of issues. I would not look at it as a gold standard for how this should behave.